haystack/components/builders/dynamic_chat_prompt_builder.py

```diff
@@ -150,17 +125,25 @@ def _process_chat_messages(self, prompt_source: List[ChatMessage], template_vari
                 f"are ChatMessage instances."
             )
 
-        last_message: ChatMessage = prompt_source[-1]
-        if last_message.is_from(ChatRole.USER):
-            template = self._validate_template(last_message.content, set(template_variables.keys()))
-            templated_user_message = ChatMessage.from_user(template.render(template_variables))
-            return prompt_source[:-1] + [templated_user_message]
-        else:
+        kwargs = kwargs or {}
+        template_variables = template_variables or {}
+        template_variables = {**kwargs, **template_variables}
+        if not template_variables:
             logger.warning(
-                "DynamicChatPromptBuilder was not provided with a user message as the last message in "
-                "chat conversation, no templating will be applied."
+                "The DynamicChatPromptBuilder run method requires template variables, but none were provided. "
+                "Please provide an appropriate template variable to enable correct prompt generation."
             )
-            return prompt_source
+        processed_messages = []
+        for message in prompt_source:
```

Comment:
@davidsbatista this is the crux of the change. Everything else is removing unnecessary internal method `_process_chat_messages` and adjusting/adding more unit tests